File: depth_first_search/go/dfs.go
package depth_first_search

type Graph struct {
	Nodes map[int][]int
}

func NewGraph() *Graph {
	return &Graph{
		Nodes: make(map[int][]int),
	}
}

func (g *Graph) AddEdge(u, v int) {
	g.Nodes[u] = append(g.Nodes[u], v)
	g.Nodes[v] = append(g.Nodes[v], u)
}

func DFS(g *Graph, v int, discovered map[int]bool, traversal *[]int) {
	discovered[v] = true
	*traversal = append(*traversal, v)

	for _, neighbour := range g.Nodes[v] {
		if _, exists := discovered[neighbour]; !exists {
			DFS(g, neighbour, discovered, traversal)
		}
	}
}

File: depth_first_search/python/dfs.py
from typing import List, Dict


class Graph:
    def __init__(self, nodes: Dict[int, List[int]]) -> None:
        self.nodes = nodes

    def __repr__(self) -> None:
        return f"Graph({self.nodes})"

    def add_edge(self, u: int, v: int) -> None:
        self.nodes[u].append(v)
        self.nodes[v].append(u)


def DFS(g: Graph, value: int, discovered: Dict[int, bool], traversal: List[int]):
    discovered[value] = True
    traversal.append(value)

    for neighbour in g.nodes[value]:
        if not discovered.get(neighbour):
            DFS(g, neighbour, discovered, traversal)
